taichi/backends/cuda/cuda_device.cpp

```diff
@@ -69,7 +71,7 @@ void CudaDevice::dealloc_memory(DeviceAllocation handle) {
       TI_ERROR("the CudaCachingAllocator is not initialized");
     }
     caching_allocator_->release(info.size, (uint64_t *)info.ptr);
-  } else {
+  } else if (!info.use_preallocated) {
```

Comment:
would you mind also explain (in code) what's the difference between `use_cached` and `use_preallocated`? ;) I feel like they might be merged into one flag.

Reply:
added in the decl part